core/src/main/java/org/apache/iceberg/ManifestReader.java

```diff
@@ -133,12 +132,17 @@ private <T extends ContentFile<T>> PartitionSpec readPartitionSpec(InputFile inp
   private static <T extends ContentFile<T>> Map<String, String> readMetadata(InputFile inputFile) {
     Map<String, String> metadata;
     try {
-      try (AvroIterable<ManifestEntry<T>> headerReader =
-          Avro.read(inputFile)
+      try (CloseableIterable<ManifestEntry<T>> headerReader =
+          InternalData.read(FileFormat.AVRO, inputFile)
               .project(ManifestEntry.getSchema(Types.StructType.of()).select("status"))
-              .classLoader(GenericManifestEntry.class.getClassLoader())
               .build()) {
-        metadata = headerReader.getMetadata();
+
+        if (headerReader instanceof AvroIterable) {
```

Comment:
We can hunt down the metadata usage separately from this PR, but this preserves existing behavior while transitioning to the new InternalData read path.

Reply:
Yeah, I think this is reasonable because it should flag cases where we are using the metadata to recover the partition spec in tests when running with Parquet or formats that don't support it.